runtime-modules/project-token/src/types.rs

```diff
@@ -0,0 +1,321 @@
+            crate::Error::<T>::InsufficientFreeBalanceForTransfer,
+        );
+        Ok(())
+    }
+
+    pub(crate) fn _total_balance(&self) -> Balance {
+        self.free_balance.saturating_add(self.reserved_balance)
+    }
+}
+/// Token Data implementation
+impl<Balance: Saturating + Copy, Hash, BlockNumber> TokenData<Balance, Hash, BlockNumber> {
+    // increase total issuance
+    pub(crate) fn increase_issuance_by(&mut self, amount: Balance) {
+        self.current_total_issuance = self.current_total_issuance.saturating_add(amount);
+    }
+
+    // decrease total issuance
+    pub(crate) fn decrease_issuance_by(&mut self, amount: Balance) {
+        self.current_total_issuance = self.current_total_issuance.saturating_sub(amount);
+    }
+}
+
+/// Encapsules parameters validation + TokenData construction
+impl<Balance: Zero + Copy + PartialOrd, Hash> TokenIssuanceParameters<Balance, Hash> {
+    /// Forward `self` state
+    pub fn try_build<T: crate::Trait, BlockNumber>(
+        self,
+        block: BlockNumber,
+    ) -> Result<TokenData<Balance, Hash, BlockNumber>, DispatchError> {
+        // validation
+
+        let patronage_info = PatronageData::<Balance, BlockNumber> {
+            last_tally_update_block: block,
+            tally: Balance::zero(),
+            rate: self.patronage_rate,
+        };
+        Ok(TokenData::<Balance, Hash, BlockNumber> {
+            current_total_issuance: self.initial_issuance,
+            issuance_state: self.initial_state,
+            transfer_policy: self.transfer_policy,
+            patronage_info,
+        })
+    }
+}
+
+impl<Hasher: Hash> MerkleProof<Hasher> {
+    pub(crate) fn verify_for_commit<T, AccountId>(
+        &self,
+        account_id: &AccountId,
+        commit: Hasher::Output,
+    ) -> DispatchResult
+    where
+        T: crate::Trait,
+        AccountId: Encode,
+    {
+        match &self.0 {
+            None => Err(crate::Error::<T>::MerkleProofNotProvided.into()),
+            Some(vec) => {
+                let init = Hasher::hash_of(account_id);
+                let proof_result = vec.iter().fold(init, |acc, (hash, side)| match side {
+                    MerkleSide::Left => Hasher::hash_of(&(hash, acc)),
+                    MerkleSide::Right => Hasher::hash_of(&(acc, hash)),
+                });
+
+                ensure!(
+                    proof_result == commit,
+                    crate::Error::<T>::MerkleProofVerificationFailure,
+                );
+
+                Ok(())
+            }
+        }
+    }
+}
+
+impl<Balance: Zero + Copy + Saturating, BlockNumber: Copy + Saturating + PartialOrd>
+    PatronageData<Balance, BlockNumber>
+{
+    pub fn outstanding_credit<BlockNumberToBalance: Convert<BlockNumber, Balance>>(
+        &self,
+        block: BlockNumber,
+    ) -> Balance {
+        let period = block.saturating_sub(self.last_tally_update_block);
+        let accrued = BlockNumberToBalance::convert(period).saturating_mul(self.rate);
```

Comment:
Should you not also multiply by issuance size during this period?

Reply:
In the vesting pallet they use `per_block: Balance` so it really is intended to be [tokens/block] as unit, that is an absolute amount of $CRT for each block.
If we want it to be dimensionless (i.e. a percentage rate) then yes the correct accounting formula would be
```
accrued = rate * period * total_supply
```
In light of latest product call the latter is a better choice

*References*: 
- https://github.com/paritytech/substrate/blob/2b4818aeb3488658ae85a5e64e9b656945553a72/frame/vesting/src/vesting_info.rs#L28

Reply:
here: cefb6b18bf4e161c0c2512591d0fe0cd805b21b4

Reply:
I suggest we make the interest rate time window configurable, see review comment.